### DISCUSSION TITLE
anonymous class parameter

### DISCUSSION THREAD

**Discussion by guoicq** [vb-consulting/Norm.net#8] [Q&A] [answered]
Let's say we have two variables.
`var id = 1; var name = "Tom";`

Right now Norm support parameters by position like below:
`connection.Execute("update table set name=@name where id = @id", name, id);`

Does Norm support anonymous class parameter?
`connection.Execute("update table set name=@name where id = @id", new {name = name, id = id});`

Or simplified version?
`connection.Execute("update table set name=@name where id = @id", new {name, id});`

By the way, for position parameters, what happen if there's repeat parameters in the query?
`connection.Execute("update table set name=@name where id = @id and parentId = @id", name, id);`

**Comment by vbilopav** [comment]
> 
> Does Norm support anonymous class parameter?
> connection.Execute("update table set name=@name where id = @id", new {name = name, id = id});
> 

Unfortunately, no at this moment.

However, currently, there are several ways to pass parameters by name. For example:

You can pass a normal, non-anonymous class instance. [Example](https://github.com/vb-consulting/Norm.net/blob/master/EXAMPLES.md#passing-class-instance-and-mapping-values-to-parameters).

Or, my favorite approach, you can pass a name-value tuple array [like this](https://github.com/vb-consulting/Norm.net/blob/master/EXAMPLES.md#mapping-parameters-by-specific-name):

```csharp
connection.Read<T>("select @p1, @p2, @p3", ("p1", 1), ("p2", 2), ("p3", 3));
```

Where p1 maps to @p1 and so on ...

But you are right, it should support anonymous class parameters. And I will be looking to add that support today and release a new version.

I'll let you know soon as it is done.

> By the way, for position parameters, what happen if there's repeat parameters in the query?
> connection.Execute("update table set name=@name where id = @id and parentId = @id", name, id);

Nothing good. Your parameters will mix.

Here is an example:

```
var (p1, p2) = connection.Read<int, int>("select @p, @p", 1, 2).Single();
Assert.Equal(1, p1);
Assert.Equal(1, p2); // @p2 mixed with @p1
```

I'll be looking to add an exception with a friendly message too.

**Comment by vbilopav** [comment, marked as answer]
I've released a new version with those changes:
[https://github.com/vb-consulting/Norm.net/blob/master/CHANGES.md#337](https://github.com/vb-consulting/Norm.net/blob/master/CHANGES.md#337)

**Reply by guoicq** [reply]
That's amazing the feature was added so quick. Thank you so much.
